### PR TITLE
Added support of auxiliary_signals in NxSpectrum

### PR DESCRIPTION
**Pull request by loichuder**

### CODE REVIEW

File: src/h5web/vis-packs/core/line/DataCurve.tsx [hunk 1]
```diff
@@ -22,58 +22,16 @@ function DataCurve(props: Props): ReactElement {
...
-      if (Number.isFinite(yErrTop)) {
-        const errTopVector = new Vector3(x, yErrTop, 0);
-        errorBarSegments.push(dataVector, errTopVector);
-        errorCapPoints.push(errTopVector);
-      }
-    });
-
-    return { data: dataPoints, bars: errorBarSegments, caps: errorCapPoints };
-  }, [abscissaScale, abscissas, camera.far, errors, ordinateScale, ordinates]);
+  const points = useCanvasPoints(abscissas, ordinates, errors);
```
Comment (loichuder) [inline src/h5web/vis-packs/core/line/DataCurve.tsx:34]: `eslint` was bugging me about the complexity of this component due to the new ternary above. I resolved it by putting all the points computation in a new hook.

File: src/h5web/vis-packs/core/line/MappedLineVis.tsx [hunk 2]
```diff
@@ -53,27 +47,25 @@ function MappedLineVis(props: Props): ReactElement {
...
-  const dataDomain = useDomain(
-    dataValues,
-    yScaleType,
-    showErrors || !autoScale ? errorValues : undefined
+  const dataDomain = useAutoScaledDomain(
+    dataArray,
+    showErrors ? errorArray : undefined,
+    auxArrays,
+    autoScale,
+    yScaleType
```
Comment (loichuder) [inline src/h5web/vis-packs/core/line/MappedLineVis.tsx:68]: Rather than making a new hook only for the `auxArrays`, I crossed the bridge and put the `autoScale` code in it as well.

Again, I welcome suggestions for the name.

File: src/h5web/vis-packs/core/line/models.ts [hunk 3]
```diff
@@ -15,3 +16,8 @@ export const GLYPH_URLS = {
   Square: squareURL,
   Cap: capURL,
 };
+
+export interface MappedArray {
+  base: ndarray | undefined;
+  mapped: ndarray | undefined;
+}
```
Comment (loichuder) [inline src/h5web/vis-packs/core/line/models.ts:23]: The new interface returned by `useArrays`.

`MappedArray` is **again** not an ideal name

File: src/h5web/providers/mock/metadata-utils.ts [hunk 4]
```diff
@@ -254,6 +265,10 @@ export function makeNxDataGroup<
     | { axes: T; axesAttr: (Extract<keyof T, string> | '.')[] }
     | { axes?: never; axesAttr?: never }
   ) &
+    (
+      | { auxiliary: T; auxAttr: (Extract<keyof T, string> | '.')[] }
```
Comment (axelboc) [inline src/h5web/providers/mock/metadata-utils.ts:269]: I don't think `'.'` should be allowed here.

File: src/h5web/vis-packs/core/hooks.ts [hunk 5]
```diff
@@ -143,3 +163,17 @@ export function useWheelCapture() {
...
+  dimMapping: DimensionMapping
+) {
+  const values = useDatasetValues(auxDatasets);
+  return useMemo(() => {
+    return Object.entries(values).map(([name, value]) => {
+      const base = getBaseArray(value, dims);
+      return { name, base, mapped: getMappedArray(base, dimMapping) };
+    });
+  }, [dimMapping, dims, values]);
+}
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/hooks.ts:179]: Yep, I get you, had the same problem with `useDatasetValues` and in other places.

File: src/h5web/vis-packs/core/line/DataCurve.tsx [hunk 6]
```diff
@@ -22,58 +22,16 @@ function DataCurve(props: Props): ReactElement {
...
+    color = '--secondary-dark',
     curveType = CurveType.LineOnly,
   } = props;
 
-  const { camera, gl } = useThree();
-  const { abscissaScale, ordinateScale } = useCanvasScales();
+  const { gl } = useThree();
+  const curveColor = color.startsWith('--')
+    ? window.getComputedStyle(gl.domElement).getPropertyValue(color).trim()
+    : color;
```
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/line/DataCurve.tsx:32]: Clever 👍

File: src/h5web/vis-packs/core/line/LineVis.tsx [hunk 7]
```diff
@@ -122,6 +124,17 @@ function LineVis(props: Props): ReactElement {
...
+          return (
+            <DataCurve
+              key={name}
+              color="--secondary"
+              curveType={curveType}
+              abscissas={abscissas}
+              ordinates={array.data as number[]}
+            />
+          );
+        })}
```
Comment (axelboc) [inline src/h5web/vis-packs/core/line/LineVis.tsx:137]: you can condense this by removing the `return`

File: src/h5web/vis-packs/nexus/containers/NxSpectrumContainer.tsx [hunk 8]
```diff
@@ -16,6 +16,7 @@ function NxSpectrumContainer(props: VisContainerProps): ReactElement {
 
   const nxData = useNxData(entity);
   const {
+    auxiliaryDatasets,
```
Comment (axelboc) [inline src/h5web/vis-packs/nexus/containers/NxSpectrumContainer.tsx:19]: Might be simpler to just name this property `auxDatasets` in the model.

**Review by axelboc — CHANGES_REQUESTED** [review]
I get how it works and all, and kudos to you, but I'm sorry to say that this feels very convoluted. I don't think you've made the right abstractions in `MappedLineVis`. Passing those `{ base, mapped }` objects makes things particularly hard to follow.

**I would remove the `useArrays` hook, inline both the `useAuxDatasetArrays` and `useAutoScaledDomain` in `MappedLineVis` for now, and reassess from there.**

I know it's going to make `MappedLineVis` unreadable, with lots of new difficult-to-name constants, but I'm sure it will help us uncover a better way of abstracting the complexity. If you agree with this approach, we can LiveShare when you're done to discuss the best way to abstract the complexity. (If you get a complexity warning, just disable it at the top of the file for now.)

File: src/h5web/vis-packs/core/utils.ts [hunk 9]
```diff
@@ -262,3 +271,65 @@ export function isScaleType(val: unknown): val is ScaleType {
...
+      Math.max(domain[1], domainToCombine[1]),
+    ],
+    otherDomains
+  );
+}
+
+export function getBaseArray<T extends unknown[] | undefined>(
+  value: T,
+  rawDims: number[]
+): T extends (infer U)[] ? ndarray<U> : undefined;
```
Comment (axelboc) [inline src/h5web/vis-packs/core/utils.ts:301]: I wonder whether we still need those override declarations, since `useMappedArray` has one already. Yeah... probably for `useMappedArrays`... Might be worth trying to remove the override declaration on `useMappedArray, actually. Or maybe you've already tried and didn't get anywhere. Not a merge blocker either way.

File: src/stories/Home.stories.mdx [hunk 10]
```diff
@@ -99,7 +99,7 @@ const dataArray = getMockDataArray('/nD_datasets/twoD');
 The library exposes a number of utility functions, which are documented below, as well as all the enums and types used by these utility functions.
 If you use only the top-level visualization components, the only utilities you need are `getDomain` and/or `useDomain`.
 
-- **`getDomain(values: number[], scaleType: ScaleType = ScaleType.Linear): Domain | undefined`** - Find the min and max values contained in a flat array of numbers. If `scaleType` is `ScaleType.Log` and the domain crosses zero, clamp the min to the first positive value or return `undefined` if the domain is not supported (i.e. `[-x, 0]`).
+- **`getDomain(values: ndarray | number[], scaleType: ScaleType = ScaleType.Linear): Domain | undefined`** - Find the min and max values contained in a array (or ndarray) of numbers. If `scaleType` is `ScaleType.Log` and the domain crosses zero, clamp the min to the first positive value or return `undefined` if the domain is not supported (i.e. `[-x, 0]`).
```
Comment (axelboc) [inline src/stories/Home.stories.mdx:102]: Maybe also change the example in the previous section to pass the `dataArray` rather than the flat values?

**Review by axelboc — CHANGES_REQUESTED** [review]
Nicely done, I think it looks great! 👏 Only small tweaks, and one question about the override typing declarations of the hooks.

File: src/h5web/vis-packs/core/line/LineVis.tsx [hunk 11]
```diff
@@ -62,6 +64,14 @@ function LineVis(props: Props): ReactElement {
     );
   }
 
+  auxArrays.forEach((auxArray) => {
+    if (auxArray.size !== dataArray.size) {
+      throw new Error(
+        `Auxiliary size (${auxArray.size}) does not match data length (${dataArray.size})`
+      );
+    }
+  });
```
[the remark this answers is not in the document]
Comment (loichuder) [reply src/h5web/vis-packs/core/line/LineVis.tsx:73]: But then I cannot display `auxArray.size` right ?

[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/line/LineVis.tsx:73]: Ah right, hmm -- I don't think it's worth displaying the size given how unlikely the error is to occur -- or if it occurs how easy it is to check the dimensions of the auxiliary arrays....

[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/line/LineVis.tsx:73]: But I don't mind either way.

[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/line/LineVis.tsx:73]: By the way, is it really a big deal if some aux arrays are shorter than the data array?

File: src/h5web/vis-packs/core/utils.ts [hunk 12]
```diff
@@ -262,3 +271,65 @@ export function isScaleType(val: unknown): val is ScaleType {
...
+
+export function getCombinedDomain(
+  domain: Domain | undefined,
+  domainsToCombine: (Domain | undefined)[]
+): Domain | undefined {
+  if (domainsToCombine.length === 0) {
+    return domain;
+  }
+
+  const [domainToCombine, ...otherDomains] = domainsToCombine;
```
[the remark this answers is not in the document]
Comment (loichuder) [reply src/h5web/vis-packs/core/utils.ts:283]: [domainToCombine, ...remainingDomains] ? :smile:

File: src/h5web/vis-packs/core/utils.ts (hunk 9 above)
Comment (loichuder) [reply src/h5web/vis-packs/core/utils.ts:301]: Removing the override of `useMappedArray` works :fishing_pole_and_fish:

File: src/stories/Home.stories.mdx (hunk 10 above)
Comment (loichuder) [reply src/stories/Home.stories.mdx:102]: Good call as I saw that it was still named `findDomain`...

File: src/h5web/vis-packs/core/line/LineVis.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (loichuder) [reply src/h5web/vis-packs/core/line/LineVis.tsx]: I don't really know.

Let's keep it this way for now to have a well-defined behaviour and let's see if we need to change this once we encounter real use cases

File: src/h5web/vis-packs/core/utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (axelboc) [reply src/h5web/vis-packs/core/utils.ts]: whatev's

File: src/h5web/vis-packs/core/utils.ts (hunk 9 above)
Comment (axelboc) [reply src/h5web/vis-packs/core/utils.ts:303]: Hmm still seems to be there, maybe you forgot to push?

File: src/h5web/vis-packs/core/hooks.ts [hunk 13]
```diff
@@ -143,3 +105,42 @@ export function useWheelCapture() {
...
+): T extends (infer U)[] ? [ndarray<U>, ndarray<U>] : [undefined, undefined];
+
+export function useMappedArray<T>(
+  value: T[] | undefined,
+  dims: number[],
+  mapping: DimensionMapping,
+  autoScale?: boolean
+) {
+  const baseArray = createMemo(getBaseArray)(value, dims);
+  const mappedArray = createMemo(applyMapping)(baseArray, mapping);
```
Comment (axelboc) [inline src/h5web/vis-packs/core/hooks.ts:125]: Ooh you're recreating a hook on every call here 😅 -- just move them out above `useCombinedDomain`

**Review by axelboc — APPROVED** [review]

File: src/h5web/vis-packs/core/utils.ts (hunk 9 above)
Comment (loichuder) [reply src/h5web/vis-packs/core/utils.ts:303]: Hm I don't know what I imagined but removing it doesn't work in fact.

It might be because we return either `[ndarray, ndarray]` or `[undefined, undefined]` which is interpreted as `(ndarray | undefined)[]` without the override.
